store/tikv/latch/latch.go

```diff
@@ -0,0 +1,162 @@
+// isEmpty is true when the waiting queue is empty after release current transaction,
+// otherwise return the front transaction in queue.
+func (l *Latch) release(startTS uint64, commitTS uint64) (isEmpty bool, front uint64) {
+	l.Lock()
+	defer l.Unlock()
+	if startTS != l.waiting[0] {
+		panic(fmt.Sprintf("invalid front ts %d, latch:%+v", startTS, l))
+	}
+	if commitTS > l.lastCommitTs {
+		l.lastCommitTs = commitTS
+	}
+	l.waiting = l.waiting[1:]
+	if len(l.waiting) == 0 {
+		isEmpty = true
+	} else {
+		front = l.waiting[0]
+		isEmpty = false
+	}
+	return
+}
+
+// Lock is the locks' information required for a transaction.
+type Lock struct {
+	// The slot IDs of the latches(keys) that a startTS must acquire before being able to processed.
+	requiredSlots []int
+	// The number of latches that the transaction has acquired.
+	acquiredCount int
+	// The number of latches whose waiting queue contains current transaction.
+	waitedCount int
+	// Current transaction's startTS.
+	startTS uint64
+}
+
+// NewLock creates a new lock.
+func NewLock(startTS uint64, requiredSlots []int) Lock {
+	return Lock{
+		requiredSlots: requiredSlots,
+		acquiredCount: 0,
+		waitedCount:   0,
+		startTS:       startTS,
+	}
+}
+
+// Latches which are used for concurrency control.
+// Each latch is indexed by a slot's ID, hence the term latch and slot are used in interchangeable,
+// but conceptually a latch is a queue, and a slot is an index to the queue
+type Latches []Latch
+
+// NewLatches create a Latches with fixed length,
+// the size will be rounded up to the power of 2.
+func NewLatches(size int) Latches {
+	powerOfTwoSize := 1 << uint(math.Ceil(math.Log2(float64(size))))
+	latches := make([]Latch, powerOfTwoSize, powerOfTwoSize)
+	return latches
+}
+
+// GenLock generates Lock for the transaction with startTS and keys.
+func (latches Latches) GenLock(startTS uint64, keys [][]byte) Lock {
+	hashes := make(map[int]bool)
+	for _, key := range keys {
+		hashes[latches.hash(key)] = true
+	}
+	slots := make([]int, 0, len(hashes))
+	for key := range hashes {
+		slots = append(slots, key)
+	}
+	sort.Ints(slots)
+	return NewLock(startTS, slots)
+}
+
+// hash return hash int for current key.
+func (latches Latches) hash(key []byte) int {
+	return int(murmur3.Sum32(key)) & (len(latches) - 1)
+}
+
+// Acquire tries to acquire the lock for a transaction.
+// It returns with timeout = true when the transaction is timeout(
+// when the lock.startTS is smaller than any key's last commitTS).
+// It returns with acquired = true when acquire success and the transaction
+// is ready to commit.
+func (latches Latches) Acquire(lock *Lock) (acquired, timeout bool) {
+	var newWait bool
+	for lock.acquiredCount < len(lock.requiredSlots) {
+		slotID := lock.requiredSlots[lock.acquiredCount]
+		acquired, timeout, newWait = latches[slotID].acquire(lock.startTS)
+		if newWait {
+			lock.waitedCount++
+		}
+		if timeout || !acquired {
+			return
```

Comment:
timeout doesn't clear the lock?

Reply:
I'd like to call `Release` outside this function.